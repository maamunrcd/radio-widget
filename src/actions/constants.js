export const LOAD_STATIONS = 'LOAD_STATIONS';
export const LOAD_STATIONS_SUCCESSFULLY = 'LOAD_STATIONS_SUCCESSFULLY';
export const LOAD_STATIONS_ERROR = 'LOAD_STATIONS_ERROR';
export const SELECT_STATION = 'SELECT_STATION';