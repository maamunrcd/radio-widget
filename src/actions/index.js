import { LOAD_STATIONS, LOAD_STATIONS_ERROR, LOAD_STATIONS_SUCCESSFULLY, SELECT_STATION } from './constants';

export function loadStations() {
    return {
        type: LOAD_STATIONS
    };
}

export function stationsLoaded(stationList) {
    return {
        type: LOAD_STATIONS_SUCCESSFULLY,
        stationList
    };
}

export function loadStationsError(error) {
    return {
        type: LOAD_STATIONS_ERROR,
        error
    };
}

export function selectStation(id) {
    return {
        type: SELECT_STATION,
        id
    };
}