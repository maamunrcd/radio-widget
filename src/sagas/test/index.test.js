import { put, takeLatest } from "redux-saga/effects";

import { LOAD_STATIONS } from "../../actions/constants";
import { loadStations, loadStationsError, stationsLoaded } from "../../actions";

import rootSaga, { fetchStationList } from "../index";
/* eslint-disable redux-saga/yield-effects */
describe("fetchStationList Saga", () => {
  let fetchStationListGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    fetchStationListGenerator = fetchStationList();

    const callDescriptor = fetchStationListGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it("should dispatch the stationsLoaded action if it requests the data successfully without URL", () => {
    const response = [
      {
        id: 1,
        title: "Putin FM",
        freq: "66,6",
      },
      {
        id: 2,
        title: "Dribbble FM",
        freq: "101,2",
      },
    ];

    const putDescriptor = fetchStationListGenerator.next(response).value;

    expect(putDescriptor).toEqual(put(stationsLoaded(response)));
  });

  it("should call the loadStationsError action if the response errors", () => {
    const response = new Error("Some error");

    const putDescriptor = fetchStationListGenerator.throw(response).value;

    expect(putDescriptor).toEqual(put(loadStationsError(response)));
  });
});

describe("rootSaga Saga", () => {
  const rootSagaData = rootSaga();
  it("should start task to watch for LOAD_REPOS action", () => {
    const takeLatestDescriptor = rootSagaData.next().value;

    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_STATIONS, fetchStationList)
    );
  });
});
