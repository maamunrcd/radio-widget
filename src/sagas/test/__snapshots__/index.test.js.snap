// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchStationList Saga should call the loadStationsError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "station.json",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`fetchStationList Saga should dispatch the stationsLoaded action if it requests the data successfully without URL 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "station.json",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
